cosmoz-page-location.html

```diff
@@ -1,238 +1,173 @@
-			params-object="{{_tempParamsObject}}"></iron-query-params>
+
+		<iron-query-params id="params"
+		></iron-query-params>
+
 	</template>
 	<script>
 		(function () {
 			'use strict';
 
+			const defaultObject = function () {
+					return {};
+				},
+				assign = Object.assign;
+
 			Polymer({
 				is: 'cosmoz-page-location',
 				properties: {
-					/**
-					 *
-					 */
 					appPath: {
 						type: String,
 						notify: true
 					},
 
-					/**
-					 *
-					 */
-					appQueryParams: {
-						type: Object,
-						notify: true
+					_appQueryString: {
+						type: String
 					},
 
-					/**
-					 */
-					routePath: {
-						type: String,
-						notify: true
+					_appHashString: {
+						type: String
 					},
 
-					/**
-					 */
-					routeQueryString: {
-						type: String
+					routePath: {
+						type: String,
+						notify: true,
+						observer: '_routeChanged'
 					},
 
-					/**
-					 *
-					 */
-					routeQueryParams: {
+					routeQuery: {
 						type: Object,
+						value: defaultObject,
 						notify: true
 					},
 
-					/**
-					 */
 					routeHash: {
-						type: String,
-						notify: true
-					},
-
-					routeHashParams: {
 						type: Object,
+						value: defaultObject,
 						notify: true
 					},
 
-					_urlQueryString: {
-						type: String
-					},
-
-					_urlHash: {
-						type: String
-					},
-
-					_tempParamsObject: {
-						type: Object
-					},
-
-					_tempParamsString: {
-						type: String
+					hashBang: {
+						type: Boolean,
+						readOnly: true
 					}
 				},
 
 				observers: [
-					'_computeRoutePathAndParams(_urlHash)',
-					'_updateHash(routePath, routeQueryString, routeHash)'
+					'_routeChanged(routeQuery.*)',
+					'_routeChanged(routeHash.*)',
+					'_appHashChanged(_appHashString)',
 				],
 
-				attached: function () {
-					this._attached = true;
+				_paramsToString(value) {
+					if (typeof value === 'string') {
+						return value;
+					}
+					return this.$.params._encodeParams(value)
+						.replace(/%3F/g, '?').replace(/%2F/g, '/').replace(/'/g, '%27');
 				},
 
-				getRouteUrl: function (routeParams, routeHashParams) {
-					var prefix = this._hashBang ? '#!' : '#';
-
-					return prefix + this._getRoute(routeParams, routeHashParams);
+				_paramsFromString(value) {
+					return this.$.params._decodeParams(value);
 				},
 
-				getAppUrl: function (appParams, routeParams, routeHashParams) {
-					var route = this._getRoute(routeParams, routeHashParams),
-						queryString;
+				_microUpdate(path, value = {}) {
+					const keys = Object.keys,
+						current = this.get(path),
+						currentKeys = keys(current);
+					let valueKeys = keys(value);
 
-					this._tempParamsObject = Object.assign({}, this.appQueryParams, appParams);
-					queryString = this._tempParamsString;
-
-					return this._getUrl(this.appPath, queryString, route);
+					currentKeys
+						.filter(key => valueKeys.indexOf(key) < 0)
+						.forEach(key => value[key] = null);
 
+					keys(value)
```

Comment:
Reuse `valueKeys`

Reply:
Can't reuse valueKeys as new keys might be added. Need to grab again latest keys.

Reply:
How? Setting a property to null won't remove it.

Reply:
It can add a key to `value`. If a hash param key has been deleted value won't have that key. In the code in `_microUpdate` we make sure that a key will be set to `null` if it is deleted.

Reply:
Ah! Then maybe it would be better to express this through code or comment? `const deletedProps = Object.keys(current).filter(key => valueKeys.indexOf(key) === -1);` 

Maybe actually chain with a forEach that explicitly sets null?